crates/astria-sequencer-relayer/src/data_availability.rs

```diff
@@ -381,27 +441,53 @@ impl CelestiaClient {
             }
         };
 
-        // get only rollup datas that can be deserialized
-        let mut rollup_datas: HashMap<_, _> = rsp
-            .blobs
-            .iter()
-            .filter_map(|blob| {
-                if let Ok(data) = SignedNamespaceData::<RollupNamespaceData>::from_bytes(&blob.data)
-                {
-                    Some((Namespace::new(blob.namespace_id), data))
-                } else {
-                    // FIXME: provide some info to identify the rollup namespace data?
-                    warn!("failed to deserialize rollup namespace data");
-                    None
-                }
-            })
+        let rollup_datas = filter_and_convert_rollup_data_blobs(
+            rsp.blobs,
+            &namespace_data.data.block_hash,
+            &verification_key,
+        );
+
+        // finally, extract the rollup txs from the rollup datas
+        let rollup_txs = rollup_datas
+            .into_iter()
+            .map(|(namespace, rollup_datas)| (namespace, rollup_datas.data.rollup_txs))
             .collect();
-        // retain rollup datas whose block hash matches the block hash of the namespaced data
-        rollup_datas
-            .retain(|_, rollup_data| namespace_data.data.block_hash == rollup_data.data.block_hash);
+        Ok(Some(SequencerBlockData {
+            block_hash: namespace_data.data.block_hash.clone(),
+            header: namespace_data.data.header.clone(),
+            last_commit: namespace_data.data.last_commit.clone(),
+            rollup_txs,
+        }))
+    }
+}
+
+/// Filters out blobs that cannot be deserialized to `SignedNamespaceData<RollupNamespaceData>`,
+/// whose block hash is not the same as the one provided, and whose signature cannot be verified
+/// with the provided verification key.
+fn filter_and_convert_rollup_data_blobs(
+    blobs: Vec<blob::Blob>,
+    block_hash: &[u8],
+    verification_key: &VerificationKey,
+) -> HashMap<Namespace, SignedNamespaceData<RollupNamespaceData>> {
+    // get only rollup datas that can be deserialized
+    let mut rollup_datas: HashMap<_, _> = blobs
+        .iter()
+        .filter_map(|blob| {
+            if let Ok(data) = SignedNamespaceData::<RollupNamespaceData>::from_bytes(&blob.data) {
+                Some((Namespace::new(blob.namespace_id), data))
+            } else {
+                // FIXME: provide some info to identify the rollup namespace data?
+                warn!("failed to deserialize rollup namespace data");
+                None
+            }
+        })
+        .collect();
 
-        // retain rollup datas that can be verified
-        rollup_datas.retain(|namespace, rollup_data| {
+    // retain rollup datas whose block hash matches the block hash of the namespaced data
+    rollup_datas.retain(|_, rollup_data| block_hash == rollup_data.data.block_hash);
```

Comment:
Should we log something about the data which shouldn't be retained?

Reply:
I don't think it's necessary, there might be a lot of these considering we write 15 sequencer blocks for 1 celestia block